### ISSUE TITLE
Make sure what we launch exists first

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/itch#1244]
Some games, instead of shipping with the recommended structure:

```
- project-name
  - project.exe
  - project_data
```

ship a tree with a version in there, something like:

```
- project-name
  - vx.y.z
    - project.exe
    - project_data
```

However, this should still be supported by the app.

### How it's supposed to work

For butler-powered games:

  * `butler apply` creates a list of paths that have disappeared when going from v1 to v2, and removes them after successful patching

For non-butler-powered-games:

  * `install/archive` calls out to `deploy`, which builds a list of files that disappeared from v1 to v2 (called "dinosaurs" cutely enough), and then removes them. 

### When it doesn't work

When it doesn't work, the install location ends up with something like:

```
- apps
  - project-name
    - v1.0.0
      - project.exe
      - project_data
    - v2.0.0
      - project.exe
      - project_data
```

and there's a chance that `v1.0.0/project.exe` gets launched instead of` v2.0.0/project.exe`.

I don't know why it would fail but it's been reported recently. (Note that cave logs would help: right click on the game tab, "View logs...", post to https://gist.github.com/ and post the gist link here).

### How deploy works

For [deploy](https://github.com/itchio/itch/blob/32d259cafe273c018f16093a187a590071de9198/src/util/deploy.ts#L56) to work, first the latest archive is extracted to a "staging" folder, so we have:

  - the downloaded .zip in `downloads/`
  - the contents of the downloaded zip in `apps/somezip.zip-stage/`
  - the old version in `apps/project-name`

Then, deploy builds a list of files present in `apps/somezip.zip-stage/` (stageFiles), and a list of files present in `apps/project-name` (destFiles).

It then computes the difference between these (dinosaurs), and eventually wipes them.

**Comment by fasterthanlime** [comment]
### A note about the future

In the future, `deploy` will only be used for installers (nsis, inno etc.)

It'll also probably be a butler command directly, because filesystem support in electron has bitten us many times before (it behaves funny with .asar files).

As for the others:

  * for butler-powered apps, `butler apply` will keep doing the right thing, as it's done for a while.
  * for zips, there'll be `extract-while-download`, which can output a list of extracted files in the end, so that we can cleanup (by comparing it to the list of files before it started).
  * for any format `unarchiver` handles, it gives us a list of files, so we can also get away without a "staging" folder.
  * for dmg, since we mount them, we can also do away with a staging folder

Installers are the only sore point here, they always have been :(

**Comment by fasterthanlime** [comment]
Seeems to happen for builds 33254, 33253, 33240 - I'll try to reproduce soon.

**Comment by fasterthanlime** [comment]
> Seeems to happen for builds 33254, 33253, 33240 - I'll try to reproduce soon.

Actually, the files on disk were okay for that case. The issue is that the game wasn't reconfigured on update. And, apparently we don't check that the candidate we're about to launch actually exists. Fun stuff!

**Comment by fasterthanlime** [comment]
Since the contents of the folder was okay, reduced the issue to "Make sure what we launch exists first", closed in 637df98d8fc1eaf8b46da93fd32c460f1a7609c9
